File: reports-service/src/main/java/com/example/client/logging/LoggingFilter.java
package com.example.client.logging;

import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;
import org.springframework.core.Ordered;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.web.server.*;
import reactor.core.publisher.Mono;


@Slf4j
public class LoggingFilter implements WebFilter, Ordered {

    @Override
    public Mono<Void> filter(ServerWebExchange serverWebExchange, WebFilterChain webFilterChain) {
        var contentLength = serverWebExchange.getRequest().getHeaders().get("Content-Length");
        MDC.put("method", serverWebExchange.getRequest().getMethod().name());
        MDC.put("uri", serverWebExchange.getRequest().getURI().getPath());

        ServerWebExchangeDecorator exchangeDecorator = new ServerWebExchangeDecorator(serverWebExchange){
            @Override
            public ServerHttpRequest getRequest() {
                return new RequestLoggingInterceptor(super.getRequest());
            }

            @Override
            public ServerHttpResponse getResponse() {
                return new ResponseLoggingInterceptor(super.getResponse());
            }
        };
        return webFilterChain.filter(exchangeDecorator)
                .doOnEach(_ignore -> {
                    if(contentLength == null || contentLength.size() == 0) {
                        log.info("Request");
                    }
                })
                .doOnSuccess(unused -> {
                    if(contentLength == null || contentLength.size() == 0) {
                        log.info("Request");
                    }
                })
                .doOnError(throwable ->{
                    if(contentLength == null || contentLength.size() == 0) {
                        log.error(throwable.getMessage());
                    }
                });
    }

    @Override
    public int getOrder() {
        return -1;
    }
}

File: task-service/src/main/kotlin/com/example/manager/logging/LoggingFilter.kt
package com.example.manager.logging

import com.example.manager.logger
import net.logstash.logback.argument.StructuredArguments.value
import org.slf4j.MDC
import org.springframework.cloud.sleuth.Span
import org.springframework.cloud.sleuth.Tracer
import org.springframework.core.Ordered
import org.springframework.http.server.reactive.ServerHttpRequest
import org.springframework.http.server.reactive.ServerHttpResponse
import org.springframework.web.server.*
import reactor.core.publisher.Mono

class LoggingFilter : WebFilter, Ordered {

    private val logger = logger<LoggingFilter>()

    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val contentLength = exchange.request.headers["Content-Length"]

        MDC.put("method", exchange.request.method?.name)
        MDC.put("uri", exchange.request.uri.path)

        val exchangeDecorator = object : ServerWebExchangeDecorator(exchange) {
            override fun getRequest(): ServerHttpRequest {
                return RequestLoggingInterceptor(super.getRequest())
            }

            override fun getResponse(): ServerHttpResponse {
                return ResponseLoggingInterceptor(super.getResponse())
            }
        }
        return chain.filter(exchangeDecorator)
                .doOnEach {
                    if(contentLength == null || contentLength.size == 0) {
                        logger.info("Request: method {}, uri {}", value("method", exchange.request.method), value("uri", exchange.request.uri))
                    }
                }
                .doOnSuccess {
                    if(contentLength == null || contentLength.size == 0) {
                        logger.info("Response: method {}, uri {}",
                                    value("method", exchangeDecorator.delegate.request.method),
                                    value("uri", exchangeDecorator.delegate.request.uri))
                    }
                }
                .doOnError {
                    logger.error("Response: method {}, uri {}",
                                 value("method", exchangeDecorator.delegate.request.method),
                                 value("uri", exchangeDecorator.delegate.request.uri))
                }
    }

    override fun getOrder(): Int {
        return -999
    }
}